libc/src/__support/FPUtil/FPBits.h

```diff
@@ -132,6 +132,70 @@ struct FPRepBase : public internal::FPLayout<fp_type> {
   static_assert((SIG_MASK & EXP_MASK & SIGN_MASK) == 0, "masks disjoint");
   static_assert((SIG_MASK | EXP_MASK | SIGN_MASK) == FP_MASK, "masks cover");
 
+protected:
+  enum class Exp : int32_t {
+    ZERO = 0,
+    // The exponent value for denormal numbers.
+    SUBNORMAL = -EXP_BIAS,
+    // The minimum exponent value for normal numbers.
+    MIN = SUBNORMAL + 1,
+    // The maximum exponent value for normal numbers.
+    MAX = EXP_BIAS,
+    // Special value all ones.
+    INF = MAX + 1,
+    // Aliases
+    BITS_ALL_ZEROES = SUBNORMAL,
+    BITS_ALL_ONES = INF,
+  };
+
+  enum class Sig : StorageType {
+    ZERO = 0,
+    ONE = 1,
+    MSB = StorageType(1) << (SIG_LEN - 1),
+    // Aliases
+    BITS_ALL_ZEROES = ZERO,
+    BITS_ALL_ONES = SIG_MASK,
+  };
+
+  template <typename T>
+  LIBC_INLINE static constexpr auto storage_cast(T value) {
+    return static_cast<StorageType>(value);
+  }
+
+  LIBC_INLINE friend constexpr Sig operator|(const Sig a, const Sig b) {
```

Comment:
Can we make this an `or` function rather than an operator, for greppability ?

Reply:
This seems a bit overkill, `Sig` is an implementation detail of `FPRep` and is not supposed to be accessed outside of it. I'd rather keep the operator as-is since they behave exactly as for the underlying enum type.

Reply:
Except that  `Sig | Sig` is not necessarily a `Sig` (see my comment below about closedness). Those semantics are  very surprising for `operator|`.

Reply:
This was clarified in the documentation (Sig is an opaque type _a la_ [std::byte](https://en.cppreference.com/w/cpp/types/byte)). I think [we're on the same page now](https://github.com/llvm/llvm-project/pull/78192#pullrequestreview-1823042276).